### PR TITLE
feat(protocol): Track transaction name changes

### PR DESCRIPTION
**Pull request by jan-auer**
Introduces `propagations` and `changes` that provides more insight into how a
transaction name was changed during the lifetime of a transaction span. Since
the transaction name is propagated to downstream SDKs as part of the Dynamic
Sampling Context (DSC), this is a relevant quality metric for sampling traces by
transaction name.

All traces where propagations occur before the final transaction name is set
may not be sampled consistently if there are rules matching on the transaction
name.

Ref https://github.com/getsentry/sentry-javascript/issues/5679

### CODE REVIEW

File: relay-general/src/protocol/transaction.rs [hunk 1]
```diff
@@ -108,6 +109,21 @@ impl IntoValue for TransactionSource {
 
 impl ProcessValue for TransactionSource {}
 
+#[derive(Clone, Debug, PartialEq, Empty, FromValue, IntoValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct TransactionNameChange {
+    /// Describes how the name of the previous transaction was determined.
+    pub source: Annotated<TransactionSource>,
+
+    /// The number of propagations from the start of the transactions to this change.
```
Comment (AbhiPrasad) [inline relay-general/src/protocol/transaction.rs]:
```suggestion
    /// The number of propagations from the start of the transaction to this change.
```

File: relay-general/src/protocol/transaction.rs [hunk 2]
```diff
@@ -108,6 +109,21 @@ impl IntoValue for TransactionSource {
 
 impl ProcessValue for TransactionSource {}
 
+#[derive(Clone, Debug, PartialEq, Empty, FromValue, IntoValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct TransactionNameChange {
+    /// Describes how the name of the previous transaction was determined.
```
Comment (AbhiPrasad) [inline relay-general/src/protocol/transaction.rs]:
```suggestion
    /// Describes how previous transaction name was determined.
```

**Review by AbhiPrasad — APPROVED** [review]

File: relay-general/src/protocol/transaction.rs [hunk 3]
```diff
@@ -123,6 +145,12 @@ pub struct TransactionInfo {
     /// This value will only be set if the transaction name was modified during event processing.
     #[metastructure(max_chars = "culprit", trim_whitespace = "true")]
     pub original: Annotated<String>,
+
+    /// A list of changes prior to the final transaction name.
+    pub changes: Annotated<Vec<Annotated<TransactionNameChange>>>,
```
Comment (jjbayer) [inline relay-general/src/protocol/transaction.rs:153]: We should make extra clear to SDK devs that this list is only for subsequent changes of the transaction name. That is, it should _not_ contain an element for the initial value of the transaction name.

File: relay-server/src/actors/processor.rs [hunk 4]
```diff
@@ -344,6 +344,94 @@ fn outcome_from_profile_error(err: relay_profiling::ProfileError) -> Outcome {
...
+        histogram(RelayHistograms::DynamicSamplingChanges) = changes.len() as u64,
+        source = source,
+        platform = platform,
+        sdk_name = sdk_name,
+        sdk_version = sdk_version,
+    );
+
+    if let Some(&total) = transaction_info.propagations.value() {
+        // If there was no change, there were no propagations that happened with a wrong name.
+        let change = last_change
```
Comment (jjbayer) [inline relay-server/src/actors/processor.rs:383]:
```suggestion
        let before_change = last_change
```

File: relay-server/src/actors/processor.rs [hunk 5]
```diff
@@ -344,6 +344,94 @@ fn outcome_from_profile_error(err: relay_profiling::ProfileError) -> Outcome {
...
+        metric!(
+            histogram(RelayHistograms::DynamicSamplingPropagationCount) = change,
+            source = source,
+            platform = platform,
+            sdk_name = sdk_name,
+            sdk_version = sdk_version,
+        );
+
+        let percentage = (change as f64) / (total as f64);
+        if percentage > 0.0 {
```
Comment (jjbayer) [inline relay-server/src/actors/processor.rs]:
```suggestion
        if total > 0 {
            let percentage = (change as f64) / (total as f64);
```

File: relay-server/src/actors/processor.rs [hunk 6]
```diff
@@ -344,6 +344,94 @@ fn outcome_from_profile_error(err: relay_profiling::ProfileError) -> Outcome {
...
+            );
+        }
+    }
+
+    if let (Some(&start), Some(&change), Some(&end)) = (
+        event.start_timestamp.value(),
+        last_change.and_then(|c| c.timestamp.value()),
+        event.timestamp.value(),
+    ) {
+        let delay_ms = (change - start).num_milliseconds().unsigned_abs();
```
Comment (jjbayer) [inline relay-server/src/actors/processor.rs]: If `change < start`, we are dealing with garbage data, so I would not emit a metric in that case.

File: relay-server/src/actors/processor.rs [hunk 7]
```diff
@@ -344,6 +344,94 @@ fn outcome_from_profile_error(err: relay_profiling::ProfileError) -> Outcome {
...
+            histogram(RelayHistograms::DynamicSamplingChangeDuration) = delay_ms,
+            source = source,
+            platform = platform,
+            sdk_name = sdk_name,
+            sdk_version = sdk_version,
+        );
+
+        let duration_ms = (end - start).num_milliseconds() as f64;
+        let percentage = (delay_ms as f64) / duration_ms;
+        if percentage > 0.0 {
```
Comment (jjbayer) [inline relay-server/src/actors/processor.rs]:
```suggestion
        if duration_ms > 0 {
            let percentage = (delay_ms as f64) / duration_ms;
```

**Review by jjbayer — APPROVED** [review]

**Pull request merged**